### PR TITLE
feat: better markdown support

### LINKED ISSUE #5361: Autolink in item description
Can we do autolinking from text in description of items like for example Github does or Telegram?

https://t.me/BLAGO_108
https://lum-a.com
https://discord.gg/XRGWVRhudj
https://twitter.com/FLAT_E_A_R_T_H
https://bastyon.com/gregory_the_lum_a
https://t.me/TheIllums
https://www.youtube.com/channel/UCDJ6CySN0ZkdufCqVEBViQw

image

https://beta.kodadot.xyz/rmrk2/gallery/13898634-b46a62619cb12c7a15-LUM-A-G-0443_514_MB_GREGORY_THE_ILLUMINATOR-00000443

Ref
---
- #5218

### PR DESCRIPTION
**Pull request by Jarsen136**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).

👇 \_ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [x] Feature
- [x] Refactoring

## Context
- [x] Closes #5361 support auto link in markdown content
- [x] Replace [vue-markdown-render](https://github.com/cloudacy/vue-markdown-render) with [vue-markdown](https://github.com/miaolz123/vue-markdown)

image

#### Before submitting pull request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [x] I've posted a screenshot of demonstrated change in this PR

#### Optional

- [ ] I've tested it at </bsx/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=Caiv9TbPz68q5dC8EcHu5xKYPRnremimGzqmEejDFNpWWLG)

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [x] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

image

image
`/rmrk2/gallery/13898634-b46a62619cb12c7a15-LUM-A-G-0443_514_MB_GREGORY_THE_ILLUMINATOR-00000443`

### CODE REVIEW

File: plugins/vueMarkdown.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import Vue from 'vue'
+import VueMarkdown from 'vue-markdown'
+Vue.use(VueMarkdown)
```
Comment (roiLeo) [inline plugins/vueMarkdown.ts:3]:
```ts
import VueMarkdown from 'vue-markdown'

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.use(VueMarkdown, {})
})
```
have you tried this?

**Review by roiLeo — COMMENTED** [review]
Doesn't Nuxt3 provide BuildIn [Markdown render](https://content.nuxtjs.org/guide/displaying/rendering/#contentrenderer)?

File: plugins/vueMarkdown.ts (hunk 1 above)
Comment (vikiival) [reply plugins/vueMarkdown.ts:3]: cc @Jarsen136

File: package.json [hunk 2]
```diff
@@ -114,7 +115,7 @@
     "vue-class-component": "^7.2.6",
     "vue-clipboard2": "^0.3.3",
     "vue-infinite-loading": "^2.4.5",
-    "vue-markdown-render": "^1.1.3",
+    "vue-markdown": "^2.2.4",
```
Comment (vikiival) [inline package.json:118]: https://github.com/miaolz123/vue-markdown/blob/254e36e8f520288110eb657c6d2eb309e711419e/package.json#L44

LOL

File: plugins/vueMarkdown.ts (hunk 1 above)
Comment (Jarsen136) [reply plugins/vueMarkdown.ts:3]: Oh, I just tried this but it does not work, unfortunately. So let's keep the current implement.

**Comment by Jarsen136** [comment]
> Doesn't Nuxt3 provide BuildIn [Markdown render](https://content.nuxtjs.org/guide/displaying/rendering/#contentrenderer)?

Sorry, I do know this package before. I'm also not sure if it could support as much custom functionality as provided by `vue-markdown` like auto-link. Whatever, in this PR we unify all the markdown content display into a common component, so in the future, if we would like to replace it with another md library, that would be easier.

**Comment by vikiival** [comment]
Screenshot 2023-03-28 at 14 37 57

**Review by vikiival — APPROVED** [review]

**Comment by yangwao** [comment]
socket security > snyk actually

**Comment by vikiival** [comment]
Screenshot 2023-03-31 at 16 56 08

Screenshot 2023-03-31 at 16 56 05

**Comment by Jarsen136** [comment]
> this one is high severity and real deal
> 
> https://security.snyk.io/vuln/SNYK-JS-MARKDOWNITKATEX-597160

Oh, I got it. Let me try to remove this package `markdown-it-katex` and find another better choice.

**Comment by Jarsen136** [comment]
> maybe try to transform markdown content with this package? https://mdxjs.com/

After some research, I found mdxjs is a good markdown content previewer to use on React, but not on Vue. It only supports `mdx` file render, but could not support converting md string content to vue element directly.
ref: https://mdx.nuxtjs.org/

**Comment by Jarsen136** [comment]
Eventually, it works as expected : )

This solution is here: https://github.com/miaolz123/vue-markdown/issues/112

For now, the snyk shows that there are three warning issue. I have overwritten these unsafe packages by the ones with safe version. So I think it's good to go now.

image

overwrite unsafe package:
image

**Comment by vikiival** [comment]
Screenshot 2023-04-02 at 22 26 04

So why is snyk failing?

**Comment by vikiival** [comment]
Also option is to fork the package and update it inside?

**Comment by Jarsen136** [comment]
> Screenshot 2023-04-02 at 22 26 04
> 
> So why is snyk failing?

I have mentioned the reason on the above comment. It's because snyk is not intelligent enough to know we have overwritten the package to the safe version. The snyk reports show that  both`highlight.js` and `markdown-it` have unsafe versions.  If you check the version at the `package.json` and `pnpm.lock`, you would find that we have overwritten the old version package.

> Also option is to fork the package and update it inside?

I think yes. I would choose the forked package `vue-markdown-v2` to solve it directly for now.

File: components/shared/Markdown.vue [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    class="content-markdown"
+    :anchor-attributes="{
+      target: '_blank',
+      rel: 'noopener noreferrer',
+    }"
+    @rendered="update" />
+</template>
+
+<script lang="ts" setup>
+import VueMarkdown from 'vue-markdown-v2'
```
[the remark this answers is not in the document]
Comment (Jarsen136) [reply components/shared/Markdown.vue:13]: Thanks. I will take a try later

[the remark this answers is not in the document]
Comment (vikiival) [reply components/shared/Markdown.vue:13]: Sounds good 🤗

**Comment by vikiival** [comment]
Hey,
since there is too much noise in this PR regarding many markdown packages. 

I am considering to close this PR and make a new one. 

WDYT @Jarsen136 @preschian ?

**Comment by Jarsen136** [comment]
> Hey,
> since there is too much noise in this PR regarding many markdown packages. 
> 
> I am considering to close this PR and make a new one. 
> 
> WDYT @Jarsen136 @preschian ?

OK, Let's close it. And I will raise a new PR soon

**Pull request closed without merging**